src/embed/connection.ts

```diff
@@ -0,0 +1,555 @@
+import type * as kv from '../kv/mod';
+import * as dag from '../dag/mod';
+import * as db from '../db/mod';
+import * as sync from '../sync/mod';
+import * as utf8 from '../utf8.js';
+import type {
+  BeginTryPullRequest,
+  BeginTryPullResponse,
+  CommitTransactionResponse,
+  HTTPRequestInfo,
+  MaybeEndTryPullResponse,
+  RebaseOpts,
+  TryPushRequest,
+} from '../repm-invoker';
+import type {JSONValue} from '../json';
+import {LogContext} from '../logger';
+import type {LogLevel} from '../logger';
+
+// TODO(arv): Use esbuild --define:TESTING=false
+
+let isTesting = false;
+
+export function setIsTesting(b: boolean): void {
+  isTesting = b;
+}
+
+export const testLog: {name: string; args: unknown[]}[] = [];
+
+export function clearTestLog(): void {
+  testLog.length = 0;
+}
+
+function logCall(name: string, ...args: unknown[]): void {
+  testLog.push({name, args});
+}
+
+type ConnectionMap = Map<
+  string,
+  {store: dag.Store; clientID: string; lc: LogContext}
+>;
+
+const connections: ConnectionMap = new Map();
+
+let transactionCounter = 0;
+
+function getConnection(dbName: string) {
+  const connection = connections.get(dbName);
+  if (!connection) {
+    throw new Error(`Database "${dbName}" is not open`);
+  }
+  return connection;
+}
+
+type Transaction = db.Write | db.Read;
+type TransactionsMap = Map<number, {txn: Transaction; lc: LogContext}>;
+const transactionsMap: TransactionsMap = new Map();
+
+function getTransaction(
+  transactionID: number,
+  map: TransactionsMap,
+): {txn: Transaction; lc: LogContext} {
+  const val = map.get(transactionID);
+  if (!val) {
+    throw new Error(`Transaction ${transactionID} is not open`);
+  }
+  return val;
+}
+
+function getWriteTransaction(
+  transactionID: number,
+  map: TransactionsMap,
+): {txn: db.Write; lc: LogContext} {
+  const {txn, lc} = getTransaction(transactionID, map);
+  if (txn instanceof db.Read) {
+    throw new Error('Transaction is read-only');
+  }
+  return {txn, lc};
+}
+
+export async function open(
+  dbName: string,
+  kvStore: kv.Store,
+  level: LogLevel,
+): Promise<string> {
+  const start = Date.now();
+  isTesting && logCall('open', dbName);
+
+  const lc = new LogContext(level).addContext('db', dbName);
+
+  const lc2 = lc.addContext('rpc', 'open');
+  lc2.debug?.('->', kvStore, level);
```

Comment:
Maybe all of this could be abstracted better. Maybe have a class and then the tests could implement the same interface with its required logging.

We could also use dynamic programming to automatically manage the logging